### ISSUE TITLE
EFTBase design discussion

### ISSUE THREAD

**Issue by dzhang314** [dzhang314/EFTBase.jl#1]
This package is an attempt to unify the various implementations of error-free transformations in Julia, including (but not limited to):

* [AccurateArithmetic.jl](https://github.com/JuliaMath/AccurateArithmetic.jl)
* [DoubleFloats.jl](https://github.com/JuliaMath/DoubleFloats.jl)
* [MultiFloats.jl](https://github.com/dzhang314/MultiFloats.jl)
* [StochasticArithmetic.jl](https://github.com/ffevotte/StochasticArithmetic.jl)

The first task is to identify what common EFT functionality can be extracted from all of these packages, and to what extent (if any) a common EFT implementation needs to be SIMD-aware.

**Comment by dzhang314** [comment]
@haampie points out in [this MultiFloats.jl issue](https://github.com/dzhang314/MultiFloats.jl/issues/10) that the use of [VectorizationBase.jl](https://github.com/chriselrod/VectorizationBase.jl) is currently tricky, because it assumes `@fastmath` can be applied everywhere. Of course, EFTs are sensitive to floating-point associativity, so `@fastmath` completely breaks them.

**Comment by haampie** [comment]
If we do this https://github.com/chriselrod/SIMDPirates.jl/issues/15#issuecomment-753922309 (drop fastmath in VectorizationBase by default) we can can pretty much generate optimal code for `+`, `-`, `*`, `/` without depending on VectorizationBase.jl. And we can even restrict T to floats when there's a FloatVec number type.

It's only getting a bit more iffy when there's non-trivial control flow (like in trig functions, potentially). In that case we'd have to do tricks like these: https://ispc.github.io/ispc.html#control-flow-example-if-statements. VectorizationBase.jl supports `ifelse(cond(x), true_val, false_val)` and does all the mask stuff, but more complicated control flow is not really supported.

**Comment by chriselrod** [comment]
Yeah, I'll drop fast flags by default. I'll do some restructuring of the package soon.

**Comment by dzhang314** [comment]
That's fantastic news @chriselrod ! I suppose that makes our job here in EFTBase much simpler.

What I'm currently imagining is that EFTBase will contain some hand-written implementations of basic EFTs (two-sum, three-sum, etc.) along with metaprogramming utilities for generating larger EFTs (e.g., 17-sum). The implementation of a transcendental function involves enough deliberate design choices (performance-precision tradeoffs) that I think it would be too "opinionated" to be useful outside the specific context of MultiFloats.jl. On the other hand, tools for generating high-precision accumulators/multipliers should be useful to anyone working with EFTs.

**Comment by ffevotte** [comment]
Yes, that's also what I had in mind.

Basic EFTs would not only be useful as base building blocks for other packages to depend on, but could also be considered part of an interface that numeric types might want to implement in order for EFT-based techniques to work with them.

Also, a question: do you think we should move this EFTBase repo into some GitHub organization up front?

**Comment by ffevotte** [comment]
@JeffreySarnoff What do you think of the idea to have a unique place for EFTs in the Julia ecosystem?

**Comment by dzhang314** [comment]
@ffevotte I'd be happy to move this repo elsewhere, but I must confess I don't know how GitHub Organizations work! Do you know how to initiate the process? Perhaps I need to ask the JuliaMath owners for contributor privileges?

By the way @chriselrod , do you have any documentation for VectorizationBase besides the README? (No worries if not -- I haven't written docs for MultiFloats yet either.) Like I said in Slack, I have no experience with explicitly vectorized programming in Julia. Even if it turns out not to be necessary for EFTBase, I would like to understand the role of packages like VectorizationBase.jl and SIMD.jl. Do they just provide the Julia equivalent of the `__m128`/`__m256`/`__m512` datatypes in C?

**Comment by ffevotte** [comment]
> I'd be happy to move this repo elsewhere, but I must confess I don't know how GitHub Organizations work! Do you know how to initiate the process? Perhaps I need to ask the JuliaMath owners for contributor privileges?

Well, I don't know much of the details myself. I think we should ask on Slack to confirm that the JuliaMath is a good place to host the repository (I think so too, but better be sure). We should also be able to find there (on Slack) a privileged JuliaMath member that would help transfer ownership of this repo to the organization.
I'm sure becoming members of the org would help with the transfer, but I'm not sure that would be absolutely necessary.

**Comment by chriselrod** [comment]
> By the way @chriselrod , do you have any documentation for VectorizationBase besides the README? (No worries if not -- I haven't written docs for MultiFloats yet either.) Like I said in Slack, I have no experience with explicitly vectorized programming in Julia. Even if it turns out not to be necessary for EFTBase, I would like to understand the role of packages like VectorizationBase.jl and SIMD.jl. Do they just provide the Julia equivalent of the `__m128`/`__m256`/`__m512` datatypes in C?

Yes. For example, look at [this godbolt](https://godbolt.org/z/eKar8h) of the `_mm512_fnmadd_pd` intrinsic. It produces:
```llvm
define dso_local <8 x double> @fnmadd(<8 x double> %0, <8 x double> %1, <8 x double> %2) local_unnamed_addr #0 !dbg !17 {
  call void @llvm.dbg.value(metadata <8 x double> %0, metadata !22, metadata !DIExpression()), !dbg !25
  call void @llvm.dbg.value(metadata <8 x double> %1, metadata !23, metadata !DIExpression()), !dbg !25
  call void @llvm.dbg.value(metadata <8 x double> %2, metadata !24, metadata !DIExpression()), !dbg !25
  %4 = fneg <8 x double> %1, !dbg !26
  %5 = tail call <8 x double> @llvm.fma.v8f64(<8 x double> %0, <8 x double> %4, <8 x double> %2) #2, !dbg !26
  ret <8 x double> %5, !dbg !27
}
```
While
```julia
julia> vx = Vec(ntuple(_ -> rand(), 8)...);

julia> @code_typed VectorizationBase.vfnmadd(vx, vx, vx)
CodeInfo(
1 ─ %1 = Base.getfield(a, :data)::NTuple{8, VecElement{Float64}}
│   %2 = VectorizationBase.llvmcall("%res = fneg nsz arcp contract afn reassoc <8 x double> %0\nret <8 x double> %res", NTuple{8, VecElement{Float64}}, Tuple{NTuple{8, VecElement{Float64}}}, %1)::NTuple{8, VecElement{Float64}}
│   %3 = Base.getfield(b, :data)::NTuple{8, VecElement{Float64}}
│   %4 = Base.getfield(c, :data)::NTuple{8, VecElement{Float64}}
│   %5 = VectorizationBase.llvmcall(("    declare <8 x double> @llvm.fmuladd.v8f64(<8 x double>, <8 x double>, <8 x double>)\n\n    define <8 x double> @entry(<8 x double>, <8 x double>, <8 x double>) alwaysinline {\n    top:\n        %res = call nsz arcp contract afn reassoc <8 x double> @llvm.fmuladd.v8f64(<8 x double> %0, <8 x double> %1, <8 x double> %2)\nret <8 x double> %res\n    }\n", "entry"), NTuple{8, VecElement{Float64}}, Tuple{NTuple{8, VecElement{Float64}}, NTuple{8, VecElement{Float64}}, NTuple{8, VecElement{Float64}}}, %2, %3, %4)::NTuple{8, VecElement{Float64}}
│   %6 = %new(Vec{8, Float64}, %5)::Vec{8, Float64}
└──      return %6
) => Vec{8, Float64}
```
Above, there are two llvmcalls, one for the `fneg` and the other for `@llvm.fmuladd`, which I use instead of `fma` because not all CPUs have FMA (although I could do a check for `VectorizationBase.FMA` to use `fma` instead), just like what the `C` intrinsic provides.

Normally, it uses llvm call to call functions from the [LLVM Language Reference](https://llvm.org/docs/LangRef.html). These have the advantage of (normally) being cross platform.
Rarely, when the desired function doesn't exist, it'll use llvm call for target-specific intrinsics instead, e.g. [here](https://github.com/chriselrod/VectorizationBase.jl/blob/743212e2a1ce7fe50a5c2d182522e9ea09e1859a/src/llvm_intrin/intrin_funcs.jl#L97) it has:
```julia
        decl64 = "declare <$W x i64> @llvm.x86.avx512.mask.pternlog.q.$(bits)(<$W x i64>, <$W x i64>, <$W x i64>, i32, i8)"
        instr64 = """
            %res = call <$W x i64> @llvm.x86.avx512.mask.pternlog.q.$(bits)(<$W x i64> %0, <$W x i64> %1, <$W x i64> %2, i32 $L, i8 -1)
            ret <$W x i64> %res
        """
```
For AVX512's ternary logic instruction. I implement them as I'm interested in trying them out, but you can file issues with specific requests.

**Comment by JeffreySarnoff** [comment]
> @JeffreySarnoff What do you think of the idea to have a unique place for EFTs in the Julia ecosystem?

I am "all in"!

**Comment by JeffreySarnoff** [comment]
My most current implementations are [errorfree.jl](https://github.com/JeffreySarnoff/ErrorfreeArithmetic.jl/blob/master/src/errorfree.jl) and [remainder.jl](https://github.com/JeffreySarnoff/ErrorfreeArithmetic.jl/blob/master/src/remainder.jl) and [leasterror.jl](https://github.com/JeffreySarnoff/ErrorfreeArithmetic.jl/blob/master/src/leasterror.jl).

I encourage you to incorporate `two_sqrt`, `two_inv`, `two_div` (found in [leasterror.jl](https://github.com/JeffreySarnoff/ErrorfreeArithmetic.jl/blob/master/src/leasterror.jl).
They are as close to error-free as is possible and highly reliable.  The two values returned are accurate, it is just that there may be residual error below the lsb of the second value. 
See the citations in that file for more detail.

After years of trying various approaches to naming these routines, I have settled on this:
`<how many values are returned>_<operation>` so
- `two_sum(a,b)` and `two_sum(a,b,c)` return two values (the most significant part and the error in the most significant part).
- `two_fma(a,b,c)` and `three_fma(a,b,c)` return two [three] values (the most significant part and the error in the most significant part [and the error in the error in the ms part]).

for functions that expect the arguments to be given in magnitude order,
`<how many values are returned>_hilo_<operation>` and
`<how many values are returned>_lohi_<operation>` are used.

**Comment by oscardssmith** [comment]
I was just reminded by a discourse post about [AccurateArrithmatic.jl](https://github.com/JuliaMath/AccurateArithmetic.jl ) This might be a good base for this package since it already has AVX support.
